src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,640 @@
+        self.inner_connection
+            .send_packet(req)
+            .await
+            .map_err(|err| RpcError::InternalError(err.to_string()))
+    }
+
+    /// Manually send ping by the send request
+    /// The inner can not start two loop, because the stream is unsafe
+    /// we need to start the loop in the client or higher level manually
+    /// WARN: this function does not support concurrent call
+    pub async fn ping(&self) -> Result<(), RpcError> {
+        self.inner_connection.ping().await
+    }
+}
+
+/// The receive stream for the client.
+/// The stream will be used to receive the response from the server,
+/// and clean outdated tasks when the task is timeout.
+struct ReceiveHeaderFuture<'a, P, F>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+    F: Future<Output = Result<RespHeader, RpcError>> + Unpin,
+{
+    /// The inner connection for the client.
+    client_inner: &'a RpcClientConnectionInner<P>,
+    /// The interval for period task.
+    interval: &'a mut Pin<Box<Interval>>,
+    /// The recv_future in the client.
+    recv_future: Pin<&'a mut F>,
+}
+
+impl<'a, P, F> ReceiveHeaderFuture<'a, P, F>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+    F: Future<Output = Result<RespHeader, RpcError>> + Unpin,
+{
+    /// Create a new receive header stream.
+    fn new(
+        client_inner: &'a RpcClientConnectionInner<P>,
+        interval: &'a mut Pin<Box<Interval>>,
+        recv_future: &'a mut F,
+    ) -> Self {
+        Self {
+            client_inner,
+            interval,
+            recv_future: Pin::new(recv_future),
+        }
+    }
+}
+
+impl<'a, P, F> Future for ReceiveHeaderFuture<'a, P, F>
+where
+    P: Packet + Clone + Send + Sync + 'static,
+    F: Future<Output = Result<RespHeader, RpcError>> + Unpin,
+{
+    type Output = Result<RespHeader, RpcError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let self_mut: &mut ReceiveHeaderFuture<'a, P, F> = self.get_mut();
+        let client_inner = self_mut.client_inner;
+
+        // clean timeout tasks
+        let mut tick_ready = false;
+        while self_mut.interval.as_mut().poll_tick(cx).is_ready() {
+            // consume all ticks in once call
+            tick_ready = true;
+        }
+        if tick_ready {
+            let packets_keeper: &mut PacketsKeeper<P> = client_inner.get_packets_keeper_mut();
+            packets_keeper.clean_timeout_tasks();
+        }
+
+        // recv header data
+        self_mut.recv_future.as_mut().poll(cx)
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+#[allow(clippy::indexing_slicing)]
+mod tests {
+    use tokio::net::TcpListener;
+
+    use super::*;
+    use std::time::Duration;
+
+    #[derive(Debug, Clone)]
+    pub struct TestPacket {
+        pub seq: u64,
+        pub op: u8,
+        pub timestamp: u64,
+    }
+
+    impl Encode for TestPacket {
+        fn encode(&self, _buf: &mut BytesMut) {}
```

Comment:
Should implement this, aotherwise the behavior is not tested

Reply:
OK, The encode logic here is mostly placed in the overall test, so it's simplified here, but I'll still add this.